tests/test_frameDict.py

```diff
@@ -0,0 +1,299 @@
+
+    def test_FrameDictOneFrameConstructor(self):
+        frameDict = ast.FrameDict(self.frame1)
+        self.assertIsInstance(frameDict, ast.FrameDict)
+        self.assertEqual(frameDict.nFrame, 1)
+        self.assertEqual(frameDict.getAllDomains(), ["FRAME1"])
+        self.assertEqual(frameDict.getIndex("frame1"), 1)  # should be case blind
+
+        with self.assertRaises(ValueError):
+            frameDict.getIndex("missingDomain")
+        with self.assertRaises(ValueError):
+            frameDict.getIndex("")
+
+        # Make sure the frame is deep copied
+        self.frame1.domain = "NEWDOMAIN"
+        self.assertEqual(frameDict.getFrame("FRAME1").domain, "FRAME1")
+        self.assertEqual(frameDict.getFrame(frameDict.BASE).domain, "FRAME1")
+        self.assertEqual(self.frame1.getRefCount(), 1)
+        self.assertEqual(self.frame1.getNObject(), self.initialNumFrames + 1)
+
+        # make sure BASE and CURRENT are available on the class and instance
+        self.assertEqual(ast.FrameDict.BASE, frameDict.BASE)
+        self.assertEqual(ast.FrameDict.CURRENT, frameDict.CURRENT)
+
+        self.checkCopy(frameDict)
+
+        indata = np.array([
+            [0.0, 0.1, -1.5],
+            [5.1, 0.0, 3.1],
+        ])
+        self.checkMappingPersistence(frameDict, indata)
+        self.checkPersistence(frameDict)
+        self.checkDict(frameDict)
+
+    def testFrameDictFrameSetConstructor(self):
+        frameSet = ast.FrameSet(self.frame1, self.zoomMap, self.frame2)
+        frameDict = ast.FrameDict(frameSet)
+
+        indata = np.array([[1.1, 2.1, 3.1], [1.2, 2.2, 3.2]])
+        predictedOut = indata * self.zoom
+        assert_allclose(frameDict.applyForward(indata), predictedOut)
+        assert_allclose(frameDict.applyInverse(predictedOut), indata)
+
+    def test_FrameDictAddFrame(self):
+        self.initialNumFrames = self.frame1.getNObject()  # may be >1 when run using pytest
+        frameDict = ast.FrameDict(self.frame1)
+
+        self.assertEqual(self.frame1.getNObject(), self.initialNumFrames + 1)
+        frameDict.addFrame(1, self.zoomMap, self.frame2)
+        self.assertEqual(frameDict.nFrame, 2)
+        self.frame2.domain = "NEWDOMAIN"
+        self.assertEqual(frameDict.getFrame("FRAME2").domain, "FRAME2")
+        self.assertEqual(frameDict.getFrame(frameDict.CURRENT).domain, "FRAME2")
+        self.assertEqual(self.frame2.getRefCount(), 1)
+        self.assertEqual(self.frame1.getNObject(), self.initialNumFrames + 2)
+
+        # make sure all objects were deep copied
+        self.frame1.domain = "newBase"
+        self.zoomMap.ident = "newMapping"
+        self.frame2.domain = "newCurrent"
+        self.assertEqual(frameDict.getFrame(frameDict.BASE).domain, "FRAME1")
+        self.assertEqual(frameDict.getFrame(frameDict.CURRENT).domain, "FRAME2")
+        self.assertEqual(frameDict.getMapping().ident, "zoomMap")
+        self.checkPersistence(frameDict)
+        self.checkDict(frameDict)
```

Comment:
Can you add a test for `addFrame` that tries to use a duplicate domain? Can it check that class invariants are preserved (in particular, all domains are unique, and `getIndex` can find any applicable frame by its domain)?